shared/src/metabase/shared/formatting/internal/date_formatters.cljc

```diff
@@ -0,0 +1,222 @@
+             "milliseconds" (concat iso-time-seconds [["."] :millisecond-ddd])}})
+
+(defn- time-format-for-options
+  "The time format is resolved as follows:
+  0. If a `:time-format` is provided directly as a [[builder]] structure, use that.
+  1. If a `:time-format` is provided as a string, look it up in [[constants/known-time-styles]], throwing if not found.
+  2. Check [[time-style-to-format]] for a supported `:time-style + :time-enabled` resolution pair.
+  3. Look up `:time-style` in [[constants/known-time-styles]].
+  4. Throw an exception, since the time style is unknown."
+  [{:keys [time-enabled time-format time-style] :as options}]
+  (or (when (string? time-format)
+        (or (get constants/known-time-styles time-format)
+            (throw (ex-info "Unknown time format" options))))
+      time-format
+      (get-in time-style-to-format [time-style time-enabled])
+      (get constants/known-time-styles time-style)
+      (throw (ex-info "Unknown time style" options))))
+
+;;; ------------------------------------- Custom formatters from options ---------------------------------------------
+;; These are cached, since the formatter is always identical for the same options.
+
+(defn- options->formatter*
+  [{:keys [date-enabled time-enabled] :as options}]
+  ;; TODO The original emits a console warning if the date-style is not in the overrides map. Reproduce that?
+  (let [date-format (when date-enabled (date-format-for-options options))
+        time-format (when time-enabled (time-format-for-options options))
+        format-list (if (and date-format time-format)
+                      (concat date-format [[", "]] time-format)
+                      ;; At most one format is given; use that one.
+                      ;; If neither is set, emit a warning and use ISO standard format.
+                      (or date-format
+                          time-format
+                          (do
+                            (log/warn "Unrecognized date/time format" options)
+                            fallback-iso-format)))]
+    (builder/->formatter format-list)))
+
+(def ^:private options->formatter-cache (atom {}))
+
+(defn options->formatter
+  "Given the options map, this reduces it to a formatter function.
+  Expects `date-style` and `time-style`, if provided, to be in the known set.
+  If they're unknown, this logs a warning and defaults to a full ISO 8601 string format.
+  If `date-style` or `time-style` are set to nil, that part will not be included.
+
+  The options and corresponding formatters are cached indefinitely, since there are generally only a few dozen
+  different sets of options, and from hundreds to many thousands of dates will be formatted in a typical session."
+  [options]
+  (-> (swap! options->formatter-cache
+             (fn [cache]
+               (if (contains? cache options)
+                 cache
+                 (assoc cache options (options->formatter* options)))))
+      (get options)))
```

Comment:
I thought about this. It would certainly be faster if it worked reliably.

The trouble is that these are usually received as JS objects, which are bad Clojure map keys since their `=` is identity, not value equality. So depending on exactly how the call is being made in JS-land, it may be a waste of memory and time to try to match the raw input `options`.

```clj
(get {#js {:foo "bar"} 7}  #js {:foo "bar"}) ; => nil

(let [k #js {:foo "bar"}]
  (get {m 7} m))  ; => 7

(= (hash #js {:foo "bar"}) (hash #js {:foo "bar"})) ; => false

;; not pictured: value semantics
```

If this is becoming a performance bottleneck, the real answer is to add new functions that take just `options` and return a `(fn [value] ...)` format function.

Reply:
It's using the normalized Clojure map of options, not the original input. (Implicitly, since this is only called from `m.s.formatting.date/format-date`.)

That's actually a precondition worth recording, which I've now done. (Even if this is in `internal` and you get what you deserve if you call someone else's internals.)